sql/inspect: use ASOF time for INSPECT

spilchen opened the pull request:
Previously, inspect issues used the current time for the AOST field instead of the timestamp at which the data was actually scanned. This change updates index consistency checks to use time thats flowed into the job. When calling INSPECT from SCRUB, we use the time specified in the SCRUB command.

Informs #148573

Release note: none
Epic: CRDB-30356

bghal reviewed (approved):
> @bghal reviewed 8 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153161)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @jeffswenson)

___
*[`pkg/sql/inspect/index_consistency_check.go` line 207 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153161#-OZdd0vi8eY6VvPU6GMZ:-OZdd0vj4JJy03K0G4Sn:bscpvlw) ([raw file](https://github.com/cockroachdb/cockroach/blob/013cc6ae781f0a3d8fede8d3fbda67c82892de2f/pkg/sql/inspect/index_consistency_check.go#L207)):*
> ```Go
> 	// Start transaction with ASOF time
> 	_, err = c.flowCtx.Cfg.DB.Executor().ExecEx(
> 		ctx, "inspect-begin-asof-txn", nil, /* txn */
> ```

Right spot for the flag label?


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153161)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @jeffswenson)

___
*[`pkg/sql/inspect/index_consistency_check.go` line 207 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153161#-OZdd0vi8eY6VvPU6GMZ:-OZeHIJx-IE7ziju_62G:b-zadw37) ([raw file](https://github.com/cockroachdb/cockroach/blob/013cc6ae781f0a3d8fede8d3fbda67c82892de2f/pkg/sql/inspect/index_consistency_check.go#L207)):*
<details><summary><i>Previously, bghal wrote…</i></summary><blockquote>

Right spot for the flag label?
</blockquote></details>

I'm not sure I follow what you mean. Did you prefer a different label for the BEGIN statement query? Can you clarify?


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153161)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @bghal, @fqazi, and @jeffswenson)

___
*[`pkg/sql/inspect/index_consistency_check.go` line 207 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153161#-OZdd0vi8eY6VvPU6GMZ:-OZeucip44YWzT3qF7PI:b-lx29z0) ([raw file](https://github.com/cockroachdb/cockroach/blob/013cc6ae781f0a3d8fede8d3fbda67c82892de2f/pkg/sql/inspect/index_consistency_check.go#L207)):*
<details><summary><i>Previously, bghal wrote…</i></summary><blockquote>

Should this be

```
ctx, "inspect-begin-asof-txn", nil  /* txn */,
```
</blockquote></details>

i see what you mean. It's odd looking, but this is the way the code formatter once it. My attempts to move the comma to the above was reverted.


<!-- Sent from Reviewable.io -->

fqazi reviewed (changes requested):
> @fqazi reviewed 7 of 8 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153161)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson)

___
*[`pkg/sql/inspect/index_consistency_check.go` line 207 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153161#-OZdd0vi8eY6VvPU6GMZ:-OZevfKm3rQJh_P-YCOS:bwprkuy) ([raw file](https://github.com/cockroachdb/cockroach/blob/013cc6ae781f0a3d8fede8d3fbda67c82892de2f/pkg/sql/inspect/index_consistency_check.go#L207)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

i see what you mean. It's odd looking, but this is the way the code formatter once it. My attempts to move the comma to the above was reverted.
</blockquote></details>

Since your passing in nil for txn, each call will make a new txn. You need to just do Cfg.DB.Txn then called SetFixedTimestamp in the txn closure. Or make sure you pass the same txn between calls.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/inspect: use ASOF time for INSPECT

Previously, inspect issues used the current time for the AOST field
instead of the timestamp at which the data was actually scanned. This
change updates index consistency checks to use time thats flowed into
the job. When calling INSPECT from SCRUB, we use the time specified in
the SCRUB command.

Informs #148573

Release note: none
Epic: CRDB-30356

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153161)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @bghal, @fqazi, and @jeffswenson)

___
*[`pkg/sql/inspect/index_consistency_check.go` line 207 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153161#-OZdd0vi8eY6VvPU6GMZ:-OZfhqglCRzRLaEt_jXu:b7eujh2) ([raw file](https://github.com/cockroachdb/cockroach/blob/013cc6ae781f0a3d8fede8d3fbda67c82892de2f/pkg/sql/inspect/index_consistency_check.go#L207)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Since your passing in nil for txn, each call will make a new txn. You need to just do Cfg.DB.Txn then called SetFixedTimestamp in the txn closure. Or make sure you pass the same txn between calls.
</blockquote></details>

Good catch. I ended up wrapping the query with an ASOF timestamp. I wasn't able to create a transaction with a fixed time because I need to use the iterator outside of the transaction. 

I also fixed the test, as it should have caught this. It wasn't properly corrupting the table before.


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
@spilchen Nice work! 
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @fqazi reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153161#-:-OZg1x-RBT8V0-8R4Ax2:b2z4sho)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jeffswenson)



<!-- Sent from Reviewable.io -->
